### PR TITLE
feat: add WASI-NN bindings to Wasm Workers Server

### LINKED ISSUE #198: Introduce WASI-NN for Machine Learning inference
### Is your feature request related to a problem? Please describe.

[WASI-NN](https://github.com/WebAssembly/wasi-nn) is a standard for inferencing in WebAssembly. From the README: 

> wasi-nn is a WASI API for performing ML inference. Its name derives from the fact that ML models are also known as neural networks (nn). ML models are typically trained using a large data set, resulting in one or more files that describe the model's weights. The model is then used to compute an "inference," e.g., the probabilities of classifying an image as a set of tags. This API is concerned initially with inference, not training.

The goal is to provide AI / ML capabilities to Wasm Workers Server opening new use cases. This feature will enable developers to add ML capabilities to their applications. Currently, the models are loaded from the Wasm module using the `Graph` primitive. Then, developers must convert the data to the `Tensor` primitive, send it to a ML backend at the host and retrieve the output. 

We will include an example so you can try and learn.

#### Configuration

Following the capability-based security approach of `wws`, workers won't be able to call WASI-NN APIs by default. You will need to configure it via config file:

```toml
name = "nn"
version = "1"

[features]
[features.wasi_nn]
allowed_backends = ["openvino"]
```

#### Language support

For the initial release, WASI-NN bindings will be available only to Rust through the [`wasi-nn` crate](https://github.com/bytecodealliance/wasi-nn). In the future, we will add support for more languages like JavaScript and Python.

#### Machine Learning Backends

The initial host implementation will rely on the [wasmtime-wasi-nn](https://github.com/bytecodealliance/wasmtime/tree/main/crates/wasi-nn) crate. This crate only supports [OpenVINO™](https://docs.openvino.ai/2023.0/home.html) as machine learning backend. However, in the future we will add new backends using different approaches.

### Tasks

- [x] Add WASI-NN bindings
- [x] Add configuration properties
- [x] Add example
- [ ] Add documentation

### PR DESCRIPTION
**Pull request by Angelmmiguel**
Introduce the new WASI-NN bindings to run Machine Learning (ML) inference in workers. It includes a new configuration parameter (`features.wasi_nn`) that allows you to set the allowed ML backends for a certain worker. Currently, it only supports `openvino` (OpenVINO) as a valid value.

In this PR you have a full working example you can test:
 
Screenshot 2023-08-21 at 13 41 33

### Changes

- [X] Add new WASI-NN bindings
- [X] Enable them via configuration
- [x] Write documentation

It closes #198

### CODE REVIEW

**Review by ereslibre — APPROVED** [review]
Despite I'm not very familiar with all the concepts here, LGTM

File: examples/rust-wasi-nn/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    } else {
+        bytes = image2tensor::convert_image_to_bytes(
+            path,
+            width,
+            height,
+            ImageTensorType::F32,
+            ColorOrder::BGR,
+        )
+        .unwrap();
+    }
```
[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply examples/rust-wasi-nn/src/main.rs:58]: I'm gonna remove this conditional. I used this one when I was testing the inference results. The `.bgr` contains the `BGR` tensor input data. In this case, you didn't need to apply any change. However, this is not longer required.

File: examples/rust-wasi-nn/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    ctx.get_output(0, &mut output_buffer)?;
+
+    Ok(sort_results(&output_buffer, labels))
+}
+
+// Sort the buffer of probabilities.
+fn sort_results(buffer: &[f32], labels: &Dataset) -> Vec<InferenceResult> {
+    let mut results: Vec<InferenceResult> = buffer
+        .iter()
+        .skip(1)
```
[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply examples/rust-wasi-nn/src/main.rs:76]: I can add a comment here but it seems to be an issue with the outputs. They right probabilities starts on index 1 in this case.

File: examples/rust-wasi-nn/src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ereslibre) [reply examples/rust-wasi-nn/src/main.rs:58]: Being an example I was just curious, not blockers :)

[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply examples/rust-wasi-nn/src/main.rs:58]: But you're right. Having this code here without any explanation may be confusing

File: docs/docs/features/machine-learning.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+title: Machine Learning inference
+---
+
+:::caution
+
+This is a feature preview. It will be available in v1.5.0
+
+:::
+
+Artificial Intelligence (AI) and Machine Learning (ML) are how topics in the community. This feature enables you to expand the capabilities of your workers by running ML models in your models. For example, you can develop an application that uses image classification or text-to-speech.
```
Comment (ereslibre) [inline docs/docs/features/machine-learning.md:11]:
```suggestion
Artificial Intelligence (AI) and Machine Learning (ML) are hot topics in the community. This feature enables you to expand the capabilities of your workers by running ML models in your models. For example, you can develop an application that uses image classification or text-to-speech.
```

File: docs/docs/features/machine-learning.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+To provide this feature, Wasm Workers Server relies on the [WASI-NN standard](https://github.com/WebAssembly/wasi-nn). This standard defines a set of APIs to send and retrieve data, and run the ML inference at the host side. The main benefits of this approach are to reuse the existing ML ecosystem (like Tensorflow and OpenVINO) and use hardware acceleration when it's available (GPUs, TPUs, etc.).
```
Comment (ereslibre) [inline docs/docs/features/machine-learning.md:13]:
```suggestion
To provide this feature, Wasm Workers Server relies on the [WASI-NN proposal](https://github.com/WebAssembly/wasi-nn). This proposal defines a set of APIs to send and retrieve data, and run the ML inference at the host side. The main benefits of this approach are to reuse the existing ML ecosystem (like Tensorflow and OpenVINO) and use hardware acceleration when it's available (GPUs, TPUs, etc.).
```

File: docs/docs/features/machine-learning.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+## Available backends
+
+A backend or ML engine is an application that parses the ML model, loads the inputs, run them and returns the output. There are multiple backends like PyTorch, [Tensorflow](https://www.tensorflow.org/) (and [Lite version]((https://www.tensorflow.org/lite))), [ONNX](https://onnxruntime.ai/) and [OpenVINO™](https://docs.openvino.ai/).
```
Comment (ereslibre) [inline docs/docs/features/machine-learning.md:17]:
```suggestion
A backend or ML engine is an application that parses the ML model, loads the inputs, runs them and returns the output. There are multiple backends like PyTorch, [Tensorflow](https://www.tensorflow.org/) (and [Lite version]((https://www.tensorflow.org/lite))), [ONNX](https://onnxruntime.ai/) and [OpenVINO™](https://docs.openvino.ai/).
```

File: docs/docs/features/machine-learning.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+You can find a [full working example in the project repository](https://github.com/vmware-labs/wasm-workers-server/tree/main/examples/rust-wasi-nn). In this example, you have a worker that returns a website to upload an image. When you upload it, a second worker retrieves the image and runs a [MobileNet](https://arxiv.org/abs/1704.04861) ML model to classify the content of the image.
+
+We recommend to check this example to get started with ML and Wasm Workers Server.
+
+<img alt="The sample application showing an image with a dog. The model predicts the image contains a 'Labrador retriever' with high confidence" src="/img/docs/features/wasi-nn.webp" style={{ width: "60%", margin: "2rem auto", display: "block" }}/>
+
+## Language compatibility
+
+| Language | Mount folders |
```
Comment (ereslibre) [inline docs/docs/features/machine-learning.md:66]:
```suggestion
| Language | Machine learning inference |
```

**Pull request merged**